### ISSUE TITLE
news ml extraction

### ISSUE THREAD

**Issue by b33lz3bubTH** [b33lz3bubTH/ml-projects#6]
Here is the complete pipeline walkthrough using **3 specific news articles** to show exactly how the data transforms from raw text to database entries.

### **The Input: 3 Raw News Articles**

1. **Article A (Positive):** *"Tata Motors reports a 15% rise in global sales for JLR, driven by strong demand in Europe. The stock surged in early trading."*
2. **Article B (Negative/Regulatory):** *"SEBI has barred the CEO of Brightcom Group from holding key managerial positions due to accounting irregularities."*
3. **Article C (Macro/Ambiguous):** *"Oil prices jumped 4% today amid rising tensions in the Middle East, raising concerns for Indian paint companies."*

---

### **Step 1: Entity Extractor**

**Job:** Identify *who* (Tickers/IDs) and *what sector* is involved. It ignores non-financial nouns.

* **Article A Output:**
```json
{
  "entities": [
    {"name": "Tata Motors", "ticker": "TATAMOTORS.NS", "type": "COMPANY"},
    {"name": "JLR", "parent": "TATAMOTORS.NS", "type": "SUBSIDIARY"}
  ]
}

```

* **Article B Output:**
```json
{
  "entities": [
    {"name": "Brightcom Group", "ticker": "BCG.NS", "type": "COMPANY"},
    {"name": "SEBI", "type": "REGULATOR"}
  ]
}

```

* **Article C Output:**
```json
{
  "entities": [
    {"name": "Asian Paints", "ticker": "ASIANPAINT.NS", "type": "INFERRED_COMPANY"},
    {"name": "Berger Paints", "ticker": "BERGEPAINT.NS", "type": "INFERRED_COMPANY"},
    {"name": "Crude Oil", "ticker": "MCX_CRUDE", "type": "COMMODITY"}
  ]
}
*Note: The extractor inferred paint companies because oil is a raw material for them.*

```

---

### **Step 2: Event Classifier**

**Job:** Label *what* happened using a standard taxonomy. This assigns the "logic" for the score.

* **Article A Output:**
```json
{
  "event_type": "EARNINGS_SALES_GROWTH",
  "category": "CORPORATE_ACTION",
  "sentiment_polarity": "POSITIVE"
}

```

* **Article B Output:**
```json
{
  "event_type": "REGULATORY_BAN",
  "category": "LEGAL_GOVERNANCE",
  "sentiment_polarity": "NEGATIVE"
}

```

* **Article C Output:**
```json
{
  "event_type": "MACRO_COMMODITY_RISE",
  "category": "EXTERNAL_RISK",
  "sentiment_polarity": "NEGATIVE_FOR_SECTOR" 
}
*Note: High oil is bad for paint stocks (input cost increases).*

```

---

### **Step 3: Impact Extractor (The 6 Variables)**

**Job:** Convert the event into numbers. We use 6 variables to calculate the final "Impact Score."

**The Variables:**

1. **Sentiment (-1 to +1):** Direction of news.
2. **Confidence (0 to 1):** How reliable is the source/extraction?
3. **Relevance (0 to 1):** Does it mention the stock directly or indirectly?
4. **Novelty (1.0 or 0.5):** Is this breaking news (1.0) or a repeat (0.5)?
5. **Severity (1 to 3):** Magnitude (Small fine vs. CEO fired).
6. **Reach (0 to 1):** Is it a niche blog or a national headline?

**Calculated Outputs:**

* **Article A (Tata Motors):**
* `Sentiment`: +0.8 (Strong sales)
* `Confidence`: 0.95 (Clear data)
* `Relevance`: 1.0 (Direct mention)
* `Novelty`: 1.0 (New report)
* `Severity`: 1.2 (Routine earnings, not a merger)
* `Reach`: 0.8 (Major financial news)
* **Final Score:** `(+0.8 * 0.95 * 1.0 * 1.2 * 0.8) = +0.73` **(Bullish)**

* **Article B (Brightcom):**
* `Sentiment`: -0.9 (Regulatory ban is very bad)
* `Confidence`: 1.0 (Official SEBI order)
* `Relevance`: 1.0 (Direct target)
* `Novelty`: 1.0 (Breaking)
* `Severity`: 2.5 (CEO banned is high severity)
* `Reach`: 0.5 (Mid-cap news)
* **Final Score:** `(-0.9 * 1.0 * 1.0 * 2.5 * 0.5) = -1.125` **(Very Bearish)**

* **Article C (Asian Paints):**
* `Sentiment`: -0.6 (Higher costs)
* `Confidence`: 0.8 (Market correlation, not direct)
* `Relevance`: 0.6 (Indirect impact via crude oil)
* `Novelty`: 0.5 (Oil has been rising all week)
* `Severity`: 1.1 (Standard volatility)
* `Reach`: 0.9 (Global news)
* **Final Score:** `(-0.6 * 0.8 * 0.6 * 0.5 * 1.1 * 0.9) = -0.14` **(Slightly Bearish)**

---

### **Step 4: Database**

**Job:** Store the final structured row for your trading engine to query.

| Timestamp | Ticker | Event_Type | Raw_Score | Decay_Factor | Final_Signal |
| --- | --- | --- | --- | --- | --- |
| 09:30:01 | **TATAMOTORS** | SALES_GROWTH | **+0.73** | 1.0 | **BUY** |
| 09:30:05 | **BCG** | REGULATORY_BAN | **-1.12** | 1.0 | **STRONG SELL** |
| 09:31:00 | **ASIANPAINT** | MACRO_COSTS | **-0.14** | 1.0 | **NEUTRAL/WEAK SELL** |

**What happens next?**
Your trading engine reads the "Final_Signal" column.

* It buys Tata Motors.
* It dumps Brightcom.
* It ignores Asian Paints (signal too weak) or adds it to a watch list.

**Comment by b33lz3bubTH** [comment]
This is a sophisticated, institutional-grade pipeline. The "secret sauce" here isn't just processing text; it is the **inference** (Step 1) and the **variable quantification** (Step 3).

To achieve this level of logic (e.g., inferring Asian Paints from Crude Oil prices), standard NLP (like Spacy or NLTK) will fail. You need a mix of **Large Language Models (LLMs)** for reasoning and **Vector Embeddings** for checking novelty.

Here is the exact technology stack and model selection you need for each step to build this pipeline.

---

### Step 1: Entity Extractor (The "Reasoning" Engine)

**The Challenge:** Extracting "Tata Motors" is easy. Inferring "Asian Paints" from "Crude Oil" requires world knowledge and logic, which standard extraction models lack.

* **Recommended Model:** **LLM with Function Calling**
* **Specific Models:** `Gemini 1.5 Pro` or `GPT-4o`.
* **Why:** These models support "JSON Mode" or "Function Calling." You can provide them with a list of tickers and a prompt like *"If a commodity is mentioned, identify Indian companies that use it as a raw material."*
* **Alternative (Open Source/Cheaper):** `Llama-3-70B` (via Groq for speed).


* **Helper Component: The Knowledge Base (RAG)**
* To ensure the LLM knows *exactly* which tickers belong to "Paint Companies," you cannot rely solely on the model's training data.
* **Solution:** Pass a context chunk (or use a Retrieval Augmented Generation system) containing a "Sector-to-Constituents" map.
* *Input to Model:* "Crude Oil is up." + [Context: Asian Paints uses Crude derivatives].
* *Output:* JSON with Ticker `ASIANPAINT.NS`.



### Step 2: Event Classifier

**The Challenge:** Mapping raw text to your specific taxonomy (`EARNINGS_SALES_GROWTH`, `REGULATORY_BAN`).

* **Recommended Model:** **Zero-Shot Classification or Fine-Tuned LLM**
* **Option A (Best for complex logic):** Continue using the **LLM from Step 1**. Since you are already sending the text to the LLM for entity extraction, ask it to classify the event in the same prompt to save cost and latency.
* **Option B (Fastest/Cheaper):** `DeBERTa-v3-large-mnli`. This is a HuggingFace model excellent at "Zero-Shot Classification." You feed it your list of event labels (Regulatory Ban, Macro Rise, etc.), and it assigns a probability to each without needing training.



### Step 3: Impact Extractor (The 6 Variables)

This is the most complex step because it requires calculation, memory (Novelty), and lookups (Reach).

#### A. Sentiment (-1 to +1)

* **Model:** **FinBERT** (`ProsusAI/finbert` on HuggingFace).
* **Why:** It is specifically trained on financial texts. It distinguishes that "Cost increased" is negative for a company, whereas generic sentiment models might just see "increase" as positive.

#### B. Novelty (1.0 vs 0.5)

* **Technique:** **Vector Similarity Search (Semantic Deduplication)**.
* **Model:** `text-embedding-3-small` (OpenAI) or `bge-m3` (Open Source).
* **Mechanism:**
1. Convert the current article into a vector (a list of numbers).
2. Query your vector database for articles from the last 24 hours.
3. If Cosine Similarity > 0.85, it is a **Repeat (0.5)**. If not, it is **Breaking (1.0)**.



#### C. Severity & Relevance & Confidence

* **Model:** **LLM (Reasoning)**.
* **Why:** "Severity" is subjective. An LLM can be prompted: *"On a scale of 1 to 3, how severe is a CEO ban compared to a generic fine?"*
* *Prompt Engineering:* You must provide "Few-Shot Examples" in the prompt so the LLM calibrates its scoring to your trading engine's expectations.



#### D. Reach

* **System:** **Domain Authority Lookup** (Not an AI Model).
* **Mechanism:** Use a predefined dictionary or API (like Moz or SimilarWeb data).
* `Bloomberg.com` = 1.0
* `UnknownBlog.com` = 0.2



### Step 4: Database (Storage & Execution)

**The Challenge:** You need high write speeds and time-series querying.

* **Recommended DB:** **TimescaleDB** (PostgreSQL extension) or **KDB+** (if you have a high budget).
* **Why:** You need to store the `Raw_Score` and query it based on time. A standard SQL database is fine for low volume, but for high-frequency news, a Time-Series database allows you to efficiently handle the "Decay Factor."

---

### Summary of the Architecture

| Step | Function | Technology/Model | Why? |
| --- | --- | --- | --- |
| **1** | **Entity** | **Gemini 1.5 Pro / GPT-4o** | Only top-tier LLMs can infer "Oil -> Asian Paints" reliably. |
| **2** | **Event** | **Same LLM as Step 1** | Efficient; classify while extracting entities. |
| **3a** | **Sentiment** | **FinBERT** | Specialized for financial nuance. |
| **3b** | **Novelty** | **Vector DB (Pinecone/Qdrant)** | Requires mathematical comparison of past news. |
| **3c** | **Severity** | **LLM Reasoning** | Requires qualitative judgement of news magnitude. |
| **4** | **Storage** | **TimescaleDB** | Best for time-series financial data. |

### The "Cost-Effective" Pipeline Suggestion

Running GPT-4/Gemini for *every* news article is expensive. Here is how to optimize:

1. **Filter First:** Use a cheap keyword filter or a small model (BERT) to discard non-financial news (sports, lifestyle).
2. **Batch Processing:** Send 5 headlines to the LLM in one prompt rather than 1 API call per headline.
3. **Hybrid Logic:** Use FinBERT (free/fast) for the sentiment score, and only use the LLM for the difficult parts (Entity Inference and Severity).

### Sample "Prompt" for Step 1 & 2 (LLM)

To make this work, your prompt is the most critical asset.

> **System:** You are a financial data engineer. Extract entities and classify events.
> **Context:** Crude Oil prices affect paint companies (Asian Paints, Berger) negatively due to input costs.
> **Taxonomy:** [REGULATORY_BAN, EARNINGS_SALES_GROWTH, MACRO_COMMODITY_RISE...]
> **Task:** Analyze the following text: "Oil prices jumped 4%..."
> **Output format:** JSON. Include inferred entities.

Would you like me to generate the **Python code for the "Novelty" checker** (Step 3) using vector embeddings? That is usually the hardest technical part to implement.
